superduperdb/ext/llm/base.py

```diff
@@ -0,0 +1,317 @@
+    openai_api_base: Optional[str] = None
+    openai_api_key: Optional[str] = None
+    model_name: str = "gpt-3.5-turbo"
+    chat: bool = True
+    system_prompt: Optional[str] = None
+    user_role: str = "user"
+    system_role: str = "system"
+
+    def __post_init__(self):
+        self.identifier = self.identifier or self.model_name
+        super().__post_init__()
+
+    def init(self):
+        try:
+            from openai import AsyncClient, OpenAI
+        except ImportError:
+            raise Exception("You must install openai with command 'pip install openai'")
+
+        params = {
+            "api_key": self.openai_api_key,
+            "base_url": self.openai_api_base,
+        }
+
+        self.client = OpenAI(**params)
+        self.aclient = AsyncClient(**params)
+        model_list = self.client.models.list()
+        model_set = sorted({model.id for model in model_list.data})
+        assert (
+            self.model_name in model_set
+        ), f"model_name {self.model_name} is not in model_set {model_set}"
+
+    def _generate(self, prompt: str, **kwargs: Any) -> str:
+        if self.chat:
+            return self._chat_generate(prompt, **kwargs)
+        else:
+            return self._prompt_generate(prompt, **kwargs)
+
+    def _batch_generate(self, prompts: List[str], **kwargs: Any) -> List[str]:
+        """
+        Use asyncio to batch generate text from a list of prompts.
+        """
+        return asyncio.run(self._async_batch_generate(prompts, **kwargs))
+
+    def _prompt_generate(self, prompt: str, **kwargs: Any) -> str:
+        """
+        Generate a completion for a given prompt with prompt format.
+        """
+        completion = self.client.completions.create(
+            model=self.model_name,
+            prompt=prompt,
+            **self.get_kwargs(
+                self.client.completions.create, kwargs, self.inference_kwargs
+            ),
+        )
+        return completion.choices[0].text
+
+    def _chat_generate(self, content: str, **kwargs: Any) -> str:
+        """
+        Generate a completion for a given prompt with chat format.
+        :param prompt: The prompt to generate a completion for.
+        :param kwargs: Any additional arguments to pass to the API.
+        """
+        messages = kwargs.get("messages", [])
+
+        if self.system_prompt:
+            messages = [
+                {"role": self.system_role, "content": self.system_prompt}
+            ] + messages
+
+        messages.append({"role": self.user_role, "content": content})
+        completion = self.client.chat.completions.create(
+            messages=messages,
+            model=self.model_name,
+            **self.get_kwargs(
+                self.client.chat.completions.create, kwargs, self.inference_kwargs
+            ),
+        )
+        return completion.choices[0].message.content
+
+    async def _async_generate(self, semaphore, prompt: str, **kwargs) -> str:
```

Comment:
Maybe now is the time to remove these `async` methods?

Reply:
This asynchronous method is used for the underlying implementation and is used for performance optimization when calling API reasoning in batches. It will not be exposed to users, which is different from the scenario we discussed before. However, if we do not use it, we can also use other methods to provide performance optimization during batch requests, otherwise it will block

Reply:
Ok, gotcha.